File: examples/src/fireworks/java/Fireworks.java
import io.github.bernhardfritz.p8g.Sketch;
import org.joml.Vector2f;
import org.joml.Vector4f;

import java.util.HashSet;
import java.util.Set;

public class Fireworks extends Sketch {

    Set<Firework> fireworks = new HashSet<>();
    Vector2f gravity = new Vector2f(0.f, 0.2f);

    @Override
    public void draw() {
        background(new float[] { 0.f, 0.f, 0.f, 25.5f});
        if (Random.randf(0.f, 1.f) < 0.1f) {
            Firework firework = new Firework(
              new Particle(
                      new Vector2f(Random.randf(0.f, width), height),
                      1.f,
                      new Vector2f(0.f, Random.randf(-10.f, -12.f)),
                      -1,
                      new Vector4f(255.f, 0.f, 0.f, 255.f)
              )
            );
            fireworks.add(firework);
        }
        for (Firework firework : fireworks) {
            firework.applyForce(gravity);
            firework.update();
            firework.draw();
        }
        fireworks.removeIf(Firework::isDone);
    }

    public static void main(String[] args) {
        Sketch sketch = new Fireworks();
        sketch.run(512, 512, "Fireworks");
    }
}

File: examples/src/fireworks/c/fireworks.c
#define STB_DS_IMPLEMENTATION
#include "stb_ds.h"
#define USING_NAMESPACE_P8G
#include "p8g.h"

#include "firework.h"

firework_t* fireworks = NULL;
vec2 gravity = { 0.f, 0.2f };

void draw() {
    background(0.f, 0.f, 0.f, 25.5f);
    if (random() < 0.1f) {
        firework_t firework = {
            .rocket = {
                .position = { random(width), height },
                .velocity = { 0.f, random(-10.f, -12.f) },
                .color = { 255.f, 0.f, 0.f, 255.f },
                .mass = 1.f,
                .lifespan = -1,
            },
            .exploded = 0,
            .particles = NULL,
        };
        arrpush(fireworks, firework);
    }
    for (int i = 0; i < arrlen(fireworks); i++) {
        firework_t* firework = &fireworks[i];
        firework_apply_force(firework, gravity);
        firework_update(firework);
        firework_draw(firework);
    }
    for (int i = arrlen(fireworks) - 1; i >= 0; i--) {
        firework_t* firework = &fireworks[i];
        if (firework_is_done(firework)) {
            firework_destroy(firework);
            arrdel(fireworks, i);
        }
    }
}

int main(void) {
    run(512, 512, "Fireworks");
    for (int i = 0; i < arrlen(fireworks); i++) {
        firework_t* firework = &fireworks[i];
        firework_destroy(firework);
    }
    arrfree(fireworks);
    return 0;
}